stan/math/fwd/fun/quad_form.hpp

```diff
@@ -0,0 +1,64 @@
+#ifndef STAN_MATH_FWD_FUN_QUAD_FORM_HPP
+#define STAN_MATH_FWD_FUN_QUAD_FORM_HPP
+
+#include <stan/math/fwd/core.hpp>
+#include <stan/math/fwd/fun/dot_product.hpp>
+#include <stan/math/fwd/fun/multiply.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Compute the quadratic form B^T A B.
+ *
+ * @tparam Ta type of elements in the square matrix
+ * @tparam Ra number of rows in the square matrix, can be Eigen::Dynamic
+ * @tparam Ca number of columns in the square matrix, can be Eigen::Dynamic
+ * @tparam Tb type of elements in the second matrix
+ * @tparam Rb number of rows in the second matrix, can be Eigen::Dynamic
+ * @tparam Cb number of columns in the second matrix, can be Eigen::Dynamic
+ *
+ * @param A square matrix
+ * @param B second matrix
+ * @return The quadratic form B^T A B, which is a symmetric matrix of size Cb
+ * (although symmetry is not guaranteed due to numerical precision).
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Ta, int Ra, int Ca, typename Tb, int Rb, int Cb,
+          require_any_fvar_t<Ta, Tb>...>
+inline Eigen::Matrix<return_type_t<Ta, Tb>, Cb, Cb> quad_form(
+    const Eigen::Matrix<Ta, Ra, Ca>& A, const Eigen::Matrix<Tb, Rb, Cb>& B) {
+  check_square("quad_form", "A", A);
+  check_multiplicable("quad_form", "A", A, "B", B);
+  return multiply(transpose(B), multiply(A, B));
+}
+
+/**
+ * Compute the quadratic form B^T A B.
+ *
+ * @tparam Ta type of elements in the square matrix
+ * @tparam Ra number of rows in the square matrix, can be Eigen::Dynamic
```

Comment:
I thought so. In reality we can't compile if we hardcode sizes because of `check_square`:
```
In file included from test/unit/math/prim/fun/quad_form_test.cpp:1:
In file included from ./stan/math/prim.hpp:10:
In file included from ./stan/math/prim/fun.hpp:238:
./stan/math/prim/fun/quad_form.hpp:31:3: error: no matching function for call to
      'check_square'
  check_square("quad_form", "A", A);
  ^~~~~~~~~~~~
test/unit/math/prim/fun/quad_form_test.cpp:22:31: note: in instantiation of
      function template specialization 'stan::math::quad_form<5, 5, 5, 5, double>'
      requested here
  matrix_d res2 = stan::math::quad_form(A, A);
                              ^
./stan/math/prim/err/check_square.hpp:21:13: note: candidate template ignored:
      could not match -1 against 5
inline void check_square(
            ^
1 error generated.
```